assets/js/modules/analytics-4/components/common/CustomDimensionsWrapper.js

```diff
@@ -0,0 +1,343 @@
+		if ( ! widgetSlug || ! newsKeyMetricsEnabled ) {
+			return null;
+		}
+
+		const { requiredCustomDimensions } = KEY_METRICS_WIDGETS[ widgetSlug ];
+
+		if (
+			! Array.isArray( requiredCustomDimensions ) ||
+			! requiredCustomDimensions.length
+		) {
+			return null;
+		}
+
+		return requiredCustomDimensions;
+	}, [ newsKeyMetricsEnabled, widgetSlug ] );
+
+	const hasCustomDimensions = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return true;
+		}
+
+		return select( MODULES_ANALYTICS_4 ).hasCustomDimensions(
+			customDimensions
+		);
+	} );
+	const isCreatingCustomDimensions = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return false;
+		}
+
+		return customDimensions.some( ( dimension ) =>
+			select( MODULES_ANALYTICS_4 ).isCreatingCustomDimension( dimension )
+		);
+	} );
+	const customDimensionsCreationErrors = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return [];
+		}
+
+		return customDimensions
+			.filter(
+				( dimension ) =>
+					!! select(
+						MODULES_ANALYTICS_4
+					).getCreateCustomDimensionError( dimension )
+			)
+			.map( ( dimension ) =>
+				select( MODULES_ANALYTICS_4 ).getCreateCustomDimensionError(
+					dimension
+				)
+			);
+	} );
+	const helpLink = useSelect( ( select ) => {
+		if ( ! customDimensionsCreationErrors.length ) {
+			return undefined;
+		}
+
+		return select( CORE_SITE ).getErrorTroubleshootingLinkURL(
+			customDimensionsCreationErrors.find(
+				isInsufficientPermissionsError
+			) || customDimensionsCreationErrors[ 0 ]
+		);
+	} );
+	const hasAnalyticsEditScope = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return undefined;
+		}
+
+		return select( CORE_USER ).hasScope( ANALYTICS_EDIT_SCOPE );
+	} );
+	const isSyncingAvailableCustomDimensions = useSelect( ( select ) => {
+		if ( ! customDimensions ) {
+			return false;
+		}
+
+		return select(
+			MODULES_ANALYTICS_4
+		).isSyncingAvailableCustomDimensions();
+	} );
+	const isNavigatingToOAuthURL = useSelect( ( select ) => {
+		const OAuthURL = select( CORE_USER ).getConnectURL( {
+			additionalScopes: [ ANALYTICS_EDIT_SCOPE ],
+			redirectURL: global.location.href,
+		} );
+
+		if ( ! OAuthURL ) {
+			return false;
+		}
+
+		return select( CORE_LOCATION ).isNavigatingTo( OAuthURL );
+	} );
+
+	const { syncAvailableCustomDimensions } =
+		useDispatch( MODULES_ANALYTICS_4 );
+	const { setValues } = useDispatch( CORE_FORMS );
+	const { setPermissionScopeError } = useDispatch( CORE_USER );
+
+	const loading =
+		isCreatingCustomDimensions ||
+		isSyncingAvailableCustomDimensions ||
+		isNavigatingToOAuthURL;
+
+	const commonErrorProps = {
+		headerText: title,
+		infoTooltip,
+	};
+
+	const handleCreateCustomDimensions = useCallback( () => {
+		if ( loading ) {
+			return;
+		}
+
+		setValues( FORM_CUSTOM_DIMENSIONS_CREATE, {
+			autoSubmit: true,
+		} );
+
+		if ( ! hasAnalyticsEditScope ) {
+			setPermissionScopeError( {
+				code: ERROR_CODE_MISSING_REQUIRED_SCOPE,
```

Comment:
If I remember correctly, this will show the insufficient permissions error popup and user will need to manually click on the button to proceed granting required permissions. This is not what we wanted to do when the user clicks on the retry button. AC says: `Clicking on Retry should attempt to create the custom dimensions again.`. So instead of showing the permissions dialog, we need to attempt re-creating custom dimensions again.

Reply:
Hi @eugene-manuilov. We're actually using the `skipModal` property here so that it doesn't show the modal and directly redirects to the OAuth flow. Let me know if that looks good, thanks!

Reply:
Ok, I missed that property. Thanks for letting me know. Anyway, you use this callback for all retry buttons, but we don't need to retry permissions for non-insufficient permissions errors. For other errors, we need to just retry the creation action.

Reply:
> Anyway, you use this callback for all retry buttons, but we don't need to retry permissions for non-insufficient permissions errors.

The permission error in this context is a little different from the permission scope error we're dispatching. The permission error here would occur if the user doesn't have access to the connected Analytics property with at least an "editor" role. However, this user will still be able to grant the necessary scope.

This is why I've dispatched the `setPermissionScopeError` for all cases as the Analytics "edit" scope is a prerequisite for creating custom dimensions.

Let me know if that sounds good, thank you!